src/blockchain_utilities/blockchain_usage.cpp

```diff
@@ -0,0 +1,248 @@
+  if (!command_line::is_arg_defaulted(vm, arg_log_level))
+    mlog_set_log(command_line::get_arg(vm, arg_log_level).c_str());
+  else
+    mlog_set_log(std::string(std::to_string(log_level) + ",bcutil:INFO").c_str());
+
+  LOG_PRINT_L0("Starting...");
+
+  bool opt_testnet = command_line::get_arg(vm, cryptonote::arg_testnet_on);
+  bool opt_rct_only = command_line::get_arg(vm, arg_rct_only);
+
+  std::string db_type = command_line::get_arg(vm, arg_database);
+  if (!cryptonote::blockchain_valid_db_type(db_type))
+  {
+    std::cerr << "Invalid database type: " << db_type << std::endl;
+    return 1;
+  }
+
+  // If we wanted to use the memory pool, we would set up a fake_core.
+
+  // Use Blockchain instead of lower-level BlockchainDB for two reasons:
+  // 1. Blockchain has the init() method for easy setup
+  // 2. exporter needs to use get_current_blockchain_height(), get_block_id_by_height(), get_block_by_hash()
+  //
+  // cannot match blockchain_storage setup above with just one line,
+  // e.g.
+  //   Blockchain* core_storage = new Blockchain(NULL);
+  // because unlike blockchain_storage constructor, which takes a pointer to
+  // tx_memory_pool, Blockchain's constructor takes tx_memory_pool object.
+  LOG_PRINT_L0("Initializing source blockchain (BlockchainDB)");
+  const std::string input = command_line::get_arg(vm, arg_input);
+  Blockchain *core_storage = NULL;
+  tx_memory_pool m_mempool(*core_storage);
+  core_storage = new Blockchain(m_mempool);
+  BlockchainDB* db = new_db(db_type);
+  if (db == NULL)
+  {
+    LOG_ERROR("Attempted to use non-existent database type: " << db_type);
+    throw std::runtime_error("Attempting to use non-existent database type");
+  }
+  LOG_PRINT_L0("database: " << db_type);
+
+  const std::string filename = input;
+  LOG_PRINT_L0("Loading blockchain from folder " << filename << " ...");
+
+  try
+  {
+    db->open(filename, DBF_RDONLY);
+  }
+  catch (const std::exception& e)
+  {
+    LOG_PRINT_L0("Error opening database: " << e.what());
+    return 1;
+  }
+  r = core_storage->init(db, opt_testnet);
+
+  CHECK_AND_ASSERT_MES(r, 1, "Failed to initialize source blockchain storage");
+  LOG_PRINT_L0("Source blockchain storage initialized OK");
+
+  LOG_PRINT_L0("Building usage patterns...");
+
+  size_t done = 0;
+  std::map<output_data, std::list<reference>> outputs;
+  std::map<uint64_t,uint64_t> indices;
+
+  LOG_PRINT_L0("Reading blockchain from " << input);
+  core_storage->for_all_transactions([&](const crypto::hash &hash, const cryptonote::transaction &tx)->bool
+  {
+    const bool coinbase = tx.vin.size() == 1 && tx.vin[0].type() == typeid(txin_gen);
+    const uint64_t height = core_storage->get_db().get_tx_block_height(hash);
+
+    // create new outputs
+    for (const auto &out: tx.vout)
+    {
+      if (opt_rct_only && out.amount)
+        continue;
+      output_data od(out.amount, indices[out.amount], coinbase, height);
+      auto it = outputs.find(od);
```

Comment:
Its not necessary to create a `output_data` object to do a find, but an `operator<` between the two types is needed. It was added in case the key was expensive to create (such as `std::string`, etc).